### PR TITLE
M5634 Security ACL framework

### PR DESCRIPTION
**Pull request by dennishendriksen**
#### Checklist
- [x] Functionality works & meets specifications
- [x] Code reviewed
- [x] Code unit/integration/system tested
- (n.a.) User documentation updated
- (n.a.) (If you have changed REST API interface) view-swagger.ftl updated
- [x] Test plan template updated
- [x] Clean commits (**squash on merge**)
- (n.a.) Added Feature/Fix to release notes
- [x] Integration tests run correctly

### CODE REVIEW

File: molgenis-bootstrap/src/main/java/org/molgenis/bootstrap/populate/UsersGroupsAuthoritiesPopulatorImpl.java [hunk 1]
```diff
@@ -94,10 +105,12 @@ public void populate()
 		Group allUsersGroup = groupFactory.create();
 		allUsersGroup.setName(AccountService.ALL_USER_GROUP);
 
-		// allow all users to update their profile
-		GroupAuthority usersGroupUserAccountAuthority = groupAuthorityFactory.create();
-		usersGroupUserAccountAuthority.setGroup(allUsersGroup);
-		usersGroupUserAccountAuthority.setRole(AUTHORITY_PLUGIN_WRITE_PREFIX + UserAccountController.ID);
+		// ----- TODO should not happen in this class
```
Comment (bartcharbon) [inline molgenis-bootstrap/src/main/java/org/molgenis/bootstrap/populate/UsersGroupsAuthoritiesPopulatorImpl.java]: resolve TODO

File: molgenis-data-plugin/src/main/java/org/molgenis/data/plugin/model/PluginSecurityRepositoryDecorator.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
+package org.molgenis.data.plugin.model;
+
+import org.molgenis.data.AbstractRepositoryDecorator;
+import org.molgenis.data.Repository;
+import org.springframework.security.acls.domain.ObjectIdentityImpl;
+import org.springframework.security.acls.model.MutableAclService;
+
+import java.util.stream.Stream;
+
+// TODO decorate all delete methods
```
Comment (bartcharbon) [inline molgenis-data-plugin/src/main/java/org/molgenis/data/plugin/model/PluginSecurityRepositoryDecorator.java]: resolve TODO

File: molgenis-app/src/main/java/org/springframework/security/acls/jdbc/AclConfig.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+import org.springframework.security.acls.model.PermissionGrantingStrategy;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+
+import javax.sql.DataSource;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * TODO move to org.molgenis sub-package once we upgraded to a spring-security-acl release with https://github.com/spring-projects/spring-security/issues/4814.
```
Comment (bartcharbon) [inline molgenis-app/src/main/java/org/springframework/security/acls/jdbc/AclConfig.java]: resolve TODO

Comment (dennishendriksen) [reply molgenis-app/src/main/java/org/springframework/security/acls/jdbc/AclConfig.java]: Can't be resolved until a future spring-security-acl artifact release.

File: molgenis-security/src/main/java/org/molgenis/security/permission/PermissionManagerServiceImpl.java [hunk 4]
```diff
@@ -79,14 +93,108 @@ public PermissionManagerServiceImpl(DataService dataService, GrantedAuthoritiesM
...
-		if (group == null) throw new RuntimeException("unknown group id [" + groupId + "]");
+		if (group == null)
+		{
+			throw new RuntimeException("unknown group id [" + groupId + "]");
+		}
+		return getPluginPermissions(group);
+	}
 
-		List<Authority> groupPermissions = getGroupPermissions(group);
-		Permissions permissions = createPermissions(groupPermissions, SecurityUtils.AUTHORITY_PLUGIN_PREFIX);
```
Comment (bartcharbon) [inline molgenis-security/src/main/java/org/molgenis/security/permission/PermissionManagerServiceImpl.java]: The switch in the createPermissions function is now obsolete, it is never called with anything else than AUTHORITY_ENTITY_PREFIX

**Comment by bartcharbon** [comment]
when saving in the permissions manager with no permissions at all (all set to none) I get an exceptions

Error! An error occurred. Please contact the administrator.
Message:Objects to lookup required

Tested with anonymous user

**Comment by bartcharbon** [comment]
Not sure which steps matter, but when I do this:
- startup with clean db 
- go straight to permissions
- click "users" to verify that "home" plugin permissions are available for anonymous
- go back to "groups", set any permission and save
- go back to "users" to see that:
anonymous has lost the home plugin permission

**Comment by fdlk** [comment]
Got NPE when assigning permissions to anonymous user
```
2018-02-07_11:03:56.938 [http-nio-8080-exec-2] ERROR o.m.c.u.a.p.PermissionManagerController - null
java.lang.NullPointerException: null
	at org.molgenis.core.ui.admin.permission.PermissionManagerController.updateUserEntityClassPermissions(PermissionManagerController.java:166)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:209)
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:136)
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:102)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:870)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:776)
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:991)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:925)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:978)
	at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:881)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:661)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:855)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:742)

```

**Comment by fdlk** [comment]
Setting permission to none in the permission manager is broken, both for EntityType and Plugin permission.

**Comment by fdlk** [comment]
Importing doesn't grant write meta permissions on imported entity type anymore (Fixed in https://github.com/dennishendriksen/molgenis/pull/37/commits/e602b743e0c4431c6a1a1b40514fba1b9c2c76f3)

File: molgenis-data-security/src/main/java/org/molgenis/data/security/meta/EntityTypeRepositorySecurityDecorator.java [hunk 5]
```diff
@@ -330,6 +333,7 @@ private void deleteEntityPermissions(String entityTypeId)
 	public void add(EntityType entity)
 	{
 		validateAddAllowed(entity);
+		createAcl(entity);
```
Comment (fdlk) [inline molgenis-data-security/src/main/java/org/molgenis/data/security/meta/EntityTypeRepositorySecurityDecorator.java]: Tricky! `validateAddAllowed` currently checks if you have the WRITE_META permission on the entity. Acl gets created after the check.

File: molgenis-data-security/src/main/java/org/molgenis/data/security/EntityTypeIdentity.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+import org.molgenis.data.meta.model.EntityType;
+import org.springframework.security.acls.domain.ObjectIdentityImpl;
+
+public class EntityTypeIdentity extends ObjectIdentityImpl
+{
+	public static final String TYPE = "entityType";
+
+	public EntityTypeIdentity(EntityType entityType)
+	{
+		this(entityType.getId());
```
Comment (fdlk) [inline molgenis-data-security/src/main/java/org/molgenis/data/security/EntityTypeIdentity.java]: Have you looked at `ObjectIdentityRetrievalStrategy`?

**Comment by fdlk** [comment]
Over ROLE_ANONYMOUS: https://docs.spring.io/spring-security/site/docs/5.0.x/reference/html/anonymous.html

**Comment by fdlk** [comment]
Issue: When I give a user permissions on the Attribute repository, the All Users group loses permissions on the Attribute repository

File: molgenis-data-security/src/main/java/org/molgenis/data/security/EntityTypeIdentity.java (hunk 6 above)
Comment (dennishendriksen) [reply molgenis-data-security/src/main/java/org/molgenis/data/security/EntityTypeIdentity.java]: Nope, I propose to look into it after merging this PR.

**Comment by tommydeboer** [comment]
Small discrepancy between old and new version. This might be fixed in the other PR, but here it is:

When I give anonymous user md_Attribute READ permission, I used to get a 'inherited from hierarchy - VIEW' indicator, but not anymore.

**Comment by tommydeboer** [comment]
*Issue*: Anonymous user doesn't inherit changes in 'All users'.

1. See that 'All users' group and Anonymous user both have READ permission on the Home plugin
2. Set the permission to EDIT for the 'All users' group
3. According to the testplan, the anonymous user now should also have EDIT permission, but it doesn't.

**Comment by dennishendriksen** [comment]
Anonymous user is not in the 'All users' group (might be an issue in the test plans).

**Comment by tommydeboer** [comment]
**Issue**: NPE when trying to view my entity in Data Explorer as user

1. Add user
2. Give user permissions 
  a. view Data Explorer and Importer
  b. view ImportRun, Data Explorer settings
3. As user
  a. Import variantsWithSamples.vcf
  c. Go to Data Explorer and open your entity

**Pull request merged**